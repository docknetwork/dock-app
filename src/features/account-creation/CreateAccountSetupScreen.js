import {
  Button,
  FormControl,
  Stack,
  Tooltip,
  Pressable,
  Popover,
} from 'native-base';
import React, {useState} from 'react';
import {useDispatch} from 'react-redux';
import {navigate} from '../../core/navigation';
import {Routes} from '../../core/routes';
import {
  Header,
  Footer,
  Content,
  ScreenContainer,
  Typography,
  SelectToggler,
  NBox as Box,
  Select,
  Input,
  InputPopover,
  LoadingButton,
} from '../../design-system';
import {BackButton} from '../../design-system/buttons';
import {
  createAccountActions,
  createAccountOperations,
} from './create-account-slice';

export function CreateAccountSetupScreen({
  form,
  onChange,
  submitDisabled,
  onSubmit,
}) {
  return (
    <ScreenContainer testID="create-wallet-screen">
      <Header>
        <BackButton />
      </Header>
      <Content marginLeft={26} marginRight={26}>
        <Typography
          fontFamily="Montserrat"
          fontSize={24}
          lineHeight={32}
          fontWeight="600"
          color="#fff"
          marginTop={52}>
          Account Name
        </Typography>
        <Typography
          fontSize={16}
          lineHeight={24}
          fontWeight="400"
          marginTop={12}>
          This will be used just for easy reference
        </Typography>
        <Box mt={7}>
          <Input
            placeholder="Account name"
            value={form.accountName}
            onChangeText={onChange('accountName')}
          />
        </Box>

        <SelectToggler placeholder="Advanced options">
          <FormControl>
            <Stack mt={7}>
              <FormControl.Label>
                Keypair crypto type
                <InputPopover>
                  Determines what cryptography will be used to create this
                  account. Note that to validate on Polkadot, the session
                  account must use "ed25519".
                </InputPopover>
              </FormControl.Label>
              <Select
                onValueChange={onChange('keypairType')}
                selectedValue={form.keypairType}>
                <Select.Item
                  label="Schnorrkel (sr25519, recommended)"
                  value="sr25519"
                />
                <Select.Item
                  label="Edwards (ed25519, alternative)"
                  value="ed25519"
                />
                <Select.Item
                  label="ECDSA (Non BTC/ETH compatible)"
                  value="ecdsa"
                />
              </Select>
            </Stack>
          </FormControl>
          <FormControl isInvalid={false}>
            <Stack mt={7}>
              <FormControl.Label>
                Secret derivation path
                <InputPopover>
                  {`You can set a custom derivation path for this account using the following syntax "/<soft-key>//<hard-key>". The "/<soft-key>" and "//<hard-key>" may be repeated and mixed. An optional "///<password>" can be used with a mnemonic seed, and may only be specified once.`}
                </InputPopover>
              </FormControl.Label>
              <Input
                placeholder="//hard/soft//password"
                onChangeText={onChange('derivationPath')}
                autoCapitalize="none"
              />
              <FormControl.ErrorMessage>
                Unable to match provided value to a secret URI.
              </FormControl.ErrorMessage>
            </Stack>
          </FormControl>
        </SelectToggler>
      </Content>
      <Footer marginBottom={114} marginLeft={26} marginRight={26}>
        <LoadingButton
          full
          testID="next-btn"
          disabled={submitDisabled}
          onPress={onSubmit}>
          Next
        </LoadingButton>
      </Footer>
    </ScreenContainer>
  );
}

export function CreateAccountSetupContainer() {
  const dispatch = useDispatch();
  const [form, setForm] = useState({
    accountName: '',
    keypairType: 'sr25519',
    derivationPath: '',
    _errors: {},
    _hasError: false,
  });

  const handleChange = key => evt => {
    setForm(v => ({
      ...v,
      [key]: evt,
    }));
  };

  const handleSubmit = () => {
    return dispatch(createAccountOperations.submitAccountForm(form));
  };

  const submitDisabled = !form.accountName;

  return (
    <CreateAccountSetupScreen
      form={form}
      submitDisabled={submitDisabled}
      onChange={handleChange}
      onSubmit={handleSubmit}
    />
  );
}
