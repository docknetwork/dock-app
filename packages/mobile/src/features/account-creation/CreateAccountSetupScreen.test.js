import React from 'react';
import {shallow} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import {CreateAccountSetupScreen} from './CreateAccountSetupScreen';

const mockStore = configureMockStore();

describe('CreateAccountSetupScreen', () => {
  it('should render correctly', () => {
    const initialState = {
      // placeholder for redux store
    };

    const wrapper = shallow(
      <CreateAccountSetupScreen
        form={{
          accountName: 'test',
        }}
        onChange={jest.fn()}
      />,
      {
        context: {store: mockStore(initialState)},
      },
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
