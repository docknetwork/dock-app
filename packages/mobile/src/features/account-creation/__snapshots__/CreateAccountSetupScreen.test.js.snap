// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccountSetupScreen should render correctly 1`] = `
<StyledNativeComponent
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "inlineStyle": InlineStyle {
        "rules": Array [
          "
  background-color: ",
          "#18181B",
          ";
  flex: 1;
",
        ],
      },
      "render": [Function],
      "shouldForwardProp": undefined,
      "styledComponentId": "StyledNativeComponent",
      "target": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
  testID="createAccountSetupScreen"
>
  <Styled(View)>
    <BackButton
      testID="createAccountSetupScreen.backBtn"
    />
  </Styled(View)>
  <Styled(ScrollView)
    marginLeft={26}
    marginRight={26}
  >
    <Styled(Component)
      marginTop={52}
      variant="h1"
    >
      Account Name
    </Styled(Component)>
    <Styled(Component)
      f={true}
      marginTop={12}
    >
      This will be used just for easy reference
    </Styled(Component)>
    <Memo(ForwardRef(Box))
      mt={7}
    >
      <Memo(ForwardRef(Input))
        placeholder="Account name"
        testID="createAccountSetupScreen.acountNameInput"
        value="test"
      />
    </Memo(ForwardRef(Box))>
    <AccountAdvancedOptions
      form={
        Object {
          "accountName": "test",
        }
      }
      onChange={
        [MockFunction] {
          "calls": Array [
            Array [
              "accountName",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
    />
  </Styled(ScrollView)>
  <Styled(Box)
    marginBottom={10}
    marginLeft={26}
    marginRight={26}
  >
    <LoadingButton
      full={true}
      testID="createAccountSetupScreen.nextBtn"
    >
      Next
    </LoadingButton>
  </Styled(Box)>
</StyledNativeComponent>
`;
